Tolerances loosened in the special multi tests.  Fixes #4749

WilkAndy opened the pull request:
These tolerances can be quite loose.  I'm changing them because on a different architecture with different compilers+lapack i get a csvdiff of 2E-3.

WilkAndy committed:
Tolerances loosened in the special multi tests.  Fixes #4749

moosebuild commented:
Results of testing 375a3e20741f4886b649e6c23ea5b7fa3066f74f using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12575

moosebuild commented:
Results of testing 375a3e20741f4886b649e6c23ea5b7fa3066f74f using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12578

moosebuild commented:
Results of testing 375a3e20741f4886b649e6c23ea5b7fa3066f74f using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12576

moosebuild commented:
Results of testing 375a3e20741f4886b649e6c23ea5b7fa3066f74f using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12577

jwpeterson commented:
In theory you should actually *tighten* your tolerances if you have architecture-dependent diffs.  Depending on the problem, a `rel_err` of 1.e-4 is already very loose, changing it to 1.e-2 makes it even worse.  As a last resort, what you can change is the tolerance that exodiff uses to determine whether a diff is real or not.

WilkAndy commented:
Hi @jwpeterson .  I've changed the "test" tolerances back to a rel_err of 1E-4.  The problem is out of my control, and must have something to do with compilers, or the BLAS being optimized differently, or precision problems, or similar.

Inside MOOSE i use PETSc's LAPACK things to calculate eigenvalues of a 3x3 matrix.  For identical 3x3 matrices i get slightly different eigenvalues on the two architectures.

I've changed a different tolerance (in the input file), so the results are less sensitive to the different eigenvalues.

@tonkmr , by the way, on this new architecture, your "tween" test fails too.  I wonder whether "tween" is also using LAPACK stuff?
linear_elastic.tween:    disp_x  rel diff: -1.1714443e-10 ~ -1.1663409e-10 = 4.35656e-03 (node 306)

moosebuild commented:
Results of testing c973f2fe83947ffa32d097b80ab604d3bb914302 using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12740

moosebuild commented:
Results of testing c973f2fe83947ffa32d097b80ab604d3bb914302 using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12743

friedmud commented:
@WilkAndy , @jwpeterson has it right.

What you need to do is _tighten_ the tolerances and then regold (create new gold files) with the tighter tolerances.  That will guard against variation from platforms...

moosebuild commented:
Results of testing c973f2fe83947ffa32d097b80ab604d3bb914302 using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12741

moosebuild commented:
Results of testing c973f2fe83947ffa32d097b80ab604d3bb914302 using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/12742

WilkAndy commented:
No, this is a classic precision-loss problem.  In my input file I have some tolerances "yield_function_tolerance" and "ep_plastic_tolerance", and i set them as low as I could initially.

With this new architecture, i get slightly different eigenvalues (and related things), differing at the last one or two significant figures - just some "random" noise.  This simply means I have to do a "random" number of different iterations before i get convergence, and i get a CSV difference.

I just needed to be less ambitious in my input-file tolerances.  I should have explained: these tolerances are identical (or analogous to) the "nl_abs_tol" quantity in MOOSE: you know if that is set too low you'll also see MOOSE never converge or take an architecture-dependent number of steps to convergence.  They are distinct from the "rel_err" in the tests file.

a

jwpeterson commented:
> these tolerances are identical (or analogous to) the "nl_abs_tol" quantity in MOOSE: you know if that is set too low you'll also see MOOSE never converge or take an architecture-dependent number of steps to convergence. They are distinct from the "rel_err" in the tests file.

Setting `nl_abs_tol` too low might cause you not to converge (assuming you also set your `nl_rel_tol` too low) but *assuming you do converge* taking an architecture-dependent number of steps because of a low `nl_abs_tol` shouldn't cause an exodiff...

Anyway, I'll trust that you understand what's going on with the tolerances in this particular application!

WilkAndy commented:
Thanks @jwpeterson.   The CSVDiff was caused because it was comparing the number of steps (the final solution for stress was indeed the same).

a
